### PR TITLE
inner_aggregation

### PR DESCRIPTION
**Pull request by michaelmattig**

### CODE REVIEW

**Comment by michaelmattig** [comment]
[Innere Aggregation bei Raster-Vektor-Join](https://trello.com/c/iqYOGYuJ/92-innere-aggregation-bei-raster-vektor-join)

File: operators/src/processing/raster_vector_join/mod.rs [hunk 1]
```diff
@@ -28,19 +29,37 @@ const MAX_NUMBER_OF_RASTER_INPUTS: usize = 8;
...
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
 pub struct RasterVectorJoinParams {
     /// Each name reflects the output column of the join result.
     /// For each raster input, one name must be defined.
     pub names: Vec<String>,
 
-    /// Specifies which method is used for aggregating values
-    pub aggregation: AggregationMethod,
+    /// Specifies which method is used for aggregating values for a feature
```
Comment (jdroenner) [inline operators/src/processing/raster_vector_join/mod.rs:38]: for each feature?

So this is the aggregation which is used to generate a single value for each feature? E.g. all pixels inside of a polygon?

File: operators/src/processing/raster_vector_join/mod.rs [hunk 2]
```diff
@@ -28,19 +29,37 @@ const MAX_NUMBER_OF_RASTER_INPUTS: usize = 8;
...
+    pub feature_aggregation: FeatureAggregationMethod,
+
+    /// Specifies which method is used for aggregating values over time
+    pub temporal_aggregation: TemporalAggregationMethod,
```
Comment (jdroenner) [inline operators/src/processing/raster_vector_join/mod.rs:42]: and this is the aggregation over time?

File: operators/src/processing/raster_vector_join/mod.rs [hunk 3]
```diff
@@ -28,19 +29,37 @@ const MAX_NUMBER_OF_RASTER_INPUTS: usize = 8;
...
+}
+
+/// How to aggreagate the values for the geometries inside a feature
+/// e.g. add all the raster values corresponding to the individual
+/// points inside a `MultiPoint` feature. Choosing None leads to an
```
Comment (jdroenner) [inline operators/src/processing/raster_vector_join/mod.rs:47]: i think we have to separate multi-features and things like polygons where each single feature can contain multiple pixels.
for points we will hit  only one pixel, however, lines and polygons can hit multiple pixels.

File: operators/src/processing/raster_vector_join/aggregator.rs [hunk 4]
```diff
@@ -1,87 +1,107 @@
...
     MeanNumber(MeanValueAggregator),
 }
 
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum AggregationResult {
+    Int(i64),
+    Float(f64),
+}
+
 impl TypedAggregator {
```
Comment (jdroenner) [inline operators/src/processing/raster_vector_join/aggregator.rs:65]: impl Aggregator for TypedAggregator will make things easier i think

File: operators/src/processing/raster_vector_join/aggregator.rs [hunk 5]
```diff
@@ -206,77 +219,74 @@ impl FirstValueOutputType for f64 {
 
 /// Aggregation function that calculates the weighted mean
 pub struct MeanValueAggregator {
-    means: Vec<f64>,
-    sum_weights: Vec<f64>,
-    null: Vec<bool>,
-    number_of_non_null_values: usize,
+    mean: f64,
+    sum_weight: f64,
+    null: bool,
```
Comment (jdroenner) [inline operators/src/processing/raster_vector_join/aggregator.rs:224]: is_null?

File: operators/src/processing/raster_vector_join/per_feature_aggregator.rs [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+}
+
+/// An aggregator wrapper for different return types
+pub enum TypedPerFeatureAggregator {
+    FirstValueFloat(FirstValueFloatPerFeatureAggregator),
+    FirstValueInt(FirstValueIntPerFeatureAggregator),
+    MeanNumber(MeanValuePerFeatureAggregator),
+}
+
+impl TypedPerFeatureAggregator {
```
Comment (jdroenner) [inline operators/src/processing/raster_vector_join/per_feature_aggregator.rs:39]: impl PerFeatureAggregator for TypedPerFeatureAggregator

**Review by jdroenner — CHANGES_REQUESTED** [review]
lets talk about this

**Pull request closed without merging**